design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,478 @@
+    name: app-redis-connection-details
+  engineVersion: "3.2"
+```
+
+The application of this RedisCluster resource claim to the API server causes
+Crossplane to provision a Redis cluster using an appropriate infrastructure
+provider. Resource claims are authored by application operators, like an OAM
+ApplicationConfig. They model the need for infrastructure from an application’s
+perspective; detailed configuration is defined by a _resource class_.
+
+Ideally resource claims (and other Kubernetes custom resources like them) could
+be first class citizens of OAM. Conversely, the contemporary OAM specification
+imposes a layer of abstraction; an extended workload type. Crossplane (as a
+tentative OAM runtime) would need to define an extended workload type
+corresponding to the above resource claim:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: ComponentSchematic
+metadata:
+  name: rediscluster
+spec:
+  workloadType: cache.crossplane.io/v1alpha1.RedisCluster
+  parameters:
+  - name: region
+    type: string
+    required: true
+  - name: engineVersion
+    type: string
+  workloadSettings:
+    - name: engineVersion
+      type: string
+      fromParam: region
+    - name: region
+      type: string
+      fromParam: region
+```
+
+In addition to defining this extended workload type, the OAM runtime would need
+to know how to render it, i.e. how to produce a RedisCluster custom resource
+given a ComponentSchematic of the RedisCluster workload type. This rendering
+process would presumably require the OAM runtime to contain logic for each
+custom resource.
+
+## Goals
+
+OAM as a specification is heavily inspired by the Kubernetes API, but is runtime
+agnostic. The purpose of this document is to explore whether the OAM
+specification could be made friendlier to - idiomatic to and compatible with -
+Kubernetes powered runtimes while remaining flexible enough to be usefully
+implemented by other runtimes.
+
+Kubernetes powered runtimes are the focus of this document because the OAM
+specification already looks and feels a lot like the Kubernetes API. When it is
+implemented using the _actual_ Kubernetes API any divergences from said API’s
+conventions are an order of magnitude more likely to violate the [principle of
+least astonishment] for users than for users of other runtimes.
+
+Any proposal outlined by this document should ensure:
+
+* The separation of concerns that is the core focus of the OAM specification.
+* That the OAM specification “look and feel” idiomatic to the Kubernetes API
+  model.
+* That the OAM specification remains useful and compelling to other runtimes.
+
+Note that this document frequently uses Crossplane as an example of a tentative
+OAM runtime that would benefit from a Kubernetes friendly OAM specification, but
+defining specifically how Crossplane will support OAM is out of scope.
+
+## Proposal
+
+This document proposes:
+
+1. An ApplicationConfiguration _references_ distinct Components and Trait
```

Comment:
is this still true? in this example https://github.com/oam-dev/spec/pull/304/files#diff-01be24953bc11e69526ef7b88a72f482R420, only scope uses the reference model.

Reply:
Good catch - it's no longer true. I'll update it.